src/modules/src/error_estimator_ukf.c

```diff
@@ -0,0 +1,1623 @@
+  }
+  return doneUpdate;
+}
+
+// reset strapdown navigation after filter update step if measurements were obtained
+static void resetNavigationStates()
+{
+  uint8_t ii;
+
+  float attVec[3] = {xEst[6], xEst[7], xEst[8]};
+  float quatNav[4] = {stateNav[6], stateNav[7], stateNav[8], stateNav[9]};
+  float attQuat[4], quatRes[4];
+
+  if ((isnan(xEst[0])) || (isnan(xEst[1])) ||
+      (isnan(xEst[2])) || (isnan(xEst[3])) ||
+      (isnan(xEst[4])) || (isnan(xEst[5])) ||
+      (isnan(xEst[6])) || (isnan(xEst[7])) ||
+      (isnan(xEst[8])))
+  {
+    for (ii = 0; ii < DIM_FILTER; ii++)
+    {
+      xEst[ii] = 0.0f;
+    }
+    nanCounterFilter++;
+  }
+  else
+  {
+    // update position and velocity state
+    for (ii = 0; ii < 6; ii++)
+    {
+      stateNav[ii] = stateNav[ii] + xEst[ii];
+    }
+
+    quatFromAtt(&attVec[0], &attQuat[0]);
+    multQuat(&quatNav[0], &attQuat[0], &quatRes[0]);
+
+    for (ii = 6; ii < 10; ii++)
+    {
+      stateNav[ii] = quatRes[ii - 6];
+    }
+
+    for (ii = 0; ii < DIM_FILTER; ii++)
+    {
+      xEst[ii] = 0.0f;
+    }
+
+    directionCosineMatrix(&quatRes[0], &dcm[0][0]);
+    transposeMatrix(&dcm[0][0], &dcmTp[0][0]);
+    computeSigmaPoints();
+  }
+}
+
+static uint8_t cholesky(float *A, float *L, uint8_t n)
+{
+  for (uint8_t i = 0; i < n; i++)
+  {
+    for (uint8_t j = 0; j < (i + 1); j++)
+    {
+      float s = 0.0f;
+      for (uint8_t k = 0; k < j; k++)
+        s += L[i * n + k] * L[j * n + k];
+      L[i * n + j] = (i == j) ? sqrtf(A[i * n + i] - s) : (1.0f / L[j * n + j] * (A[i * n + j] - s));
+    }
+  }
+  return 1;
+}
+
+static void transposeMatrix(float *mat, float *matTp)
+{
+  matTp[0 * 3 + 0] = mat[0];
+  matTp[0 * 3 + 1] = mat[1 * 3 + 0];
+  matTp[0 * 3 + 2] = mat[2 * 3 + 0];
+
+  matTp[1 * 3 + 0] = mat[0 * 3 + 1];
+  matTp[1 * 3 + 1] = mat[1 * 3 + 1];
+  matTp[1 * 3 + 2] = mat[2 * 3 + 1];
+
+  matTp[2 * 3 + 0] = mat[0 * 3 + 2];
+  matTp[2 * 3 + 1] = mat[1 * 3 + 2];
+  matTp[2 * 3 + 2] = mat[2 * 3 + 2];
+}
+
+// compute direction cosine matrix to transfer between lab and body frames
+static void directionCosineMatrix(float *quat, float *dcm)
+{
+  float q0_2 = quat[0] * quat[0];
+  float q1_2 = quat[1] * quat[1];
+  float q2_2 = quat[2] * quat[2];
+  float q3_2 = quat[3] * quat[3];
+
+  // compute elements of the direction cosine matrix, ned to body
+  dcm[0 * 3 + 0] = q0_2 + q1_2 - q2_2 - q3_2;
+  dcm[0 * 3 + 1] = 2 * (quat[1] * quat[2] + quat[0] * quat[3]);
+  dcm[0 * 3 + 2] = 2 * (quat[1] * quat[3] - quat[0] * quat[2]);
+
+  dcm[1 * 3 + 0] = 2 * (quat[1] * quat[2] - quat[0] * quat[3]);
+  dcm[1 * 3 + 1] = q0_2 - q1_2 + q2_2 - q3_2;
+  dcm[1 * 3 + 2] = 2 * (quat[2] * quat[3] + quat[0] * quat[1]);
+
+  dcm[2 * 3 + 0] = 2 * (quat[1] * quat[3] + quat[0] * quat[2]);
+  dcm[2 * 3 + 1] = 2 * (quat[2] * quat[3] - quat[0] * quat[1]);
+  dcm[2 * 3 + 2] = q0_2 - q1_2 - q2_2 + q3_2;
+}
+
+// transform quaternion into euler angles
+static void quatToEuler(float *quat, float *eulerAngles)
```

Comment:
Would it be possible to use the standard math functions that are existing in the firmware? This for instance:

https://github.com/bitcraze/crazyflie-firmware/blob/af96c070d6bc32cf73eeb89980a970d7e01f7f8c/src/modules/interface/math3d.h#L679-L681